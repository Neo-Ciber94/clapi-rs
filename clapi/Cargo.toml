[package]
name = "clapi"
description="A framework for create command-line applications"
keywords=["command-line", "command", "framework", "parse", "parser", "parsing", "macros", "api"]
version = "0.1.0"
authors = ["Freddy L"]
edition = "2018"

[features]
macros = ["clapi_macros"]
typing = []

[dependencies]
clapi_macros = { optional=true, path="../clapi_macros" }
clapi_internal = { path= "../clapi_internal" }
serde = { version="1.0.118", features=["derive"], optional=true }

[dev-dependencies]
serde_test = "1.0.118"
serde_json = "1.0.60"

[package.metadata.docs.rs]
all-features = true
rustflags = ["--cfg", "doc_cfg"]