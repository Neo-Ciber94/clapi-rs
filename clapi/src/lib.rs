//! # Clapi
//!
//! Clapi (Command-Line API) is a framework for create command line applications.
//!
//! Currently clapi provides several methods for create command line applications:
//! - Parsing the arguments
//! - Function handlers
//! - Macros attributes
//!
//! ## Parsing the arguments
//! ```no_run
//! // TODO
//! ```
//!
//! ## Function handlers
//! ```no_run
//! // TODO
//! ```
//!
//! ## Macro attributes
//! Requires `macros` feature enable.
//!
//! ```ignore
//! #[!cfg(features="macros")]
//! use clapi::macros::*;
//!
//! // We need to mark the app entry point as a `command`
//! #[command(version=1.0)]
//! fn main(){ }
//!
//! // Mark a function as a `subcommand` and defines if `option` and `arg`
//! #[subcommand]
//! #[option(times, alias="t", default=1)]
//! #[arg(values)]
//! fn repeat(times: u32, values: Vec<String>){
//!     let values = values.join(" ");
//!     for _ in 0..times {
//!         println!("{}", values);
//!     }
//! }
//! ```


/// Utility methods and extensions intended for internal use.
#[macro_use]
pub mod utils;

mod command;
pub use command::*;

mod option;
pub use option::*;

mod args;
pub use args::*;

mod arg_count;
pub use arg_count::*;

mod command_line;
pub use command_line::*;

mod context;
pub use context::*;

mod error;
pub use error::*;

mod tokenizer;
pub use tokenizer::*;

mod parser;
pub use parser::*;

mod parse_result;
pub use parse_result::*;

mod suggestion;
pub use suggestion::*;

mod help;
pub use help::*;

mod symbol;
pub use symbol::*;

#[macro_use]
mod app_macros;
pub use app_macros::*;

#[cfg(feature="macros")]
pub mod macros {
    extern crate clapi_macros;
    pub use clapi_macros::*;
}

#[cfg(feature="macros")]
pub use macros::*;

#[doc(hidden)]
pub mod macro_utils {
    pub extern crate clapi_macros_utils;

    #[macro_export]
    macro_rules! declare_option_var {
        ($options:ident, $name:ident: $ty:ty) => {
            $crate::macro_utils::clapi_macros_utils::__declare_option_var!($options, $name: $ty)
        };
    }

    #[macro_export]
    macro_rules! declare_argument_var {
        ($arguments:ident, $name:ident: $ty:ty) => {
            $crate::macro_utils::clapi_macros_utils::__declare_argument_var!($arguments, $name: $ty)
        };
    }

    #[macro_export]
    macro_rules! var_type {
        ($ty:ty) => {
            $crate::macro_utils::clapi_macros_utils::__var_type!($ty)
        };
    }
}

#[doc(hidden)]
pub use macro_utils::*;