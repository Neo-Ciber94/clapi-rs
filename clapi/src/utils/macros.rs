/// Asserts that the given value is not empty by calling `is_empty` and returns the value.
#[macro_export]
macro_rules! assert_not_empty {
    ($value:expr) => {{
        let value = $value;
        assert!(!value.is_empty(), "`{}` is empty", stringify!(value));
        value
    }};

    ($value:expr, $args:tt) => {{
        let value = $value;
        assert!(!value.is_empty(), $args);
        value
    }};
}

/// Asserts the given `String` or `str` is not empty or consists only of whitespaces.
#[macro_export]
macro_rules! assert_not_blank {
    ($value:expr) => {{
        let value = $value;
        assert!(!value.trim().is_empty(), "`{}` is blank", stringify!(value));
        value
    }};

    ($value:expr, $args:tt) => {{
        let value = $value;
        assert!(!value.trim().is_empty(), $args);
        value
    }};
}

/// Asserts the given `String` or `str` is no empty or contains whitespaces.
#[macro_export]
macro_rules! assert_contains_no_whitespaces {
    ($value:expr) => {{
        let value = &$value;
        assert!(!value.trim().is_empty(), "`{}` is empty", stringify!($value));
        assert!(value.trim().chars().all(|c| !c.is_whitespace()), "`{}` must not contains whitespaces", stringify!($value));
    }};
}

#[cfg(test)]
mod tests {
    #[test]
    fn assert_not_empty_test1() {
        assert_not_empty!(String::from("Hello World"));
        assert_not_empty!("Hello World");
    }

    #[test]
    #[should_panic]
    fn assert_not_empty_test2() {
        let _value = assert_not_empty!(String::new());
    }

    #[test]
    #[should_panic]
    fn assert_not_empty_test3() {
        let _value = assert_not_empty!(String::from(""));
    }

    #[test]
    fn assert_not_blank_test1() {
        assert_not_blank!("Hello");
        assert_not_blank!("Hello World");
        assert_not_blank!("   a");
    }

    #[test]
    #[should_panic]
    fn assert_not_blank_test2() {
        assert_not_blank!("");
    }

    #[test]
    #[should_panic]
    fn assert_not_blank_test3() {
        assert_not_blank!("  ");
    }

    #[test]
    #[should_panic]
    fn assert_not_blank_test4() {
        assert_not_blank!("\t");
    }

    #[test]
    #[should_panic]
    fn assert_not_blank_test5() {
        assert_not_blank!("\n");
    }

    #[test]
    #[should_panic(expected="`name` must not contains whitespaces")]
    fn assert_contains_not_whitespaces_test1(){
        let name = "Hello World";
        let _ = assert_contains_no_whitespaces!(name);
    }

    #[test]
    #[should_panic(expected="`\" \"` is empty")]
    fn assert_contains_not_whitespaces_test2(){
        let _ = assert_contains_no_whitespaces!(" ");
    }
}
