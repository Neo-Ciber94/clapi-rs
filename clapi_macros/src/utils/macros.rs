
/// Quote the result of an expression
macro_rules! quote_expr {
    ($value:expr) => {{
        let val = &$value;
        quote::quote!(#val)
    }};
}