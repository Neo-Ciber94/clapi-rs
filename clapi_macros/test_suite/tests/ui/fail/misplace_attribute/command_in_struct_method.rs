use clapi::macros::*;

struct MyStruct;
impl MyStruct {
    #[command]
    fn app(&self) {}
}

fn main(){}