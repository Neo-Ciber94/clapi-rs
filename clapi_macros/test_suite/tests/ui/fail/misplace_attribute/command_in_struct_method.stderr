error: custom attribute panicked
 --> $DIR/command_in_struct_method.rs:5:5
  |
5 |     #[command]
  |     ^^^^^^^^^^
  |
  = help: message: `app` is not a top function.
          `command`s must be free functions and be declared outside a module.
